ThoughtAnnotation: calculate position based on selection range

ethan-james opened the pull request:
Fixes #3352 

As discussed, getting the bounding rect of a [selection range](https://github.com/ethan-james/em/blob/1700e10d50f6805897c77fc5f9af733ca5ab5bcc/src/components/ThoughtAnnotation.tsx#L342) seems to work well for positioning the thought annotation without duplicating the thought in an invisible container.

TODO:

- ~~get the positioning right for superscripts, urls, email~~
- ~~get the positioning right for the end-of-thought faux caret~~
- ~~create a separate selection to position the start-of-thought faux caret correctly~~
- ~~get editable's bounding rect for empty thoughts with annotations~~

---

I remember that we had an issue when we were building the faux caret where the end-of-thought position was being reported inaccurately by iOS Safari. This may also impact the positioning of superscript, which would mean we need to update some snapshots. It most likely won't create a visual discrepancy in the same way that replacing the real caret with the faux caret often does.

raineorshine commented on src/components/BulletCursorOverlay.tsx:
```diff
@@ -195,7 +195,7 @@ export default function BulletCursorOverlay({
             isTableCol1={isTableCol1}
           />
         )}
-        <ThoughtAnnotationWrapper cursorOverlay />
```
@ethan-james Can you explain this change? The `cursorOverlay` prop is used to set the positioner to overlay only. Since the cursor overlay is not used to render an actual thought, it can skip a lot of internal processing and effects. 

If this prop is removed here in `BulletCursorOverlay`, I'm not sure where else it is used. 

Please discuss with @cindmichelle who is responsible for the cursor overlay implementation. Once you come to a consensus, I can check in. Thanks!

raineorshine commented on src/components/ThoughtAnnotation.tsx:
```diff
@@ -223,6 +214,11 @@ const ThoughtAnnotationContainer = React.memo(
 
+    // We're trying to get rid of contentWidth as part of #3369, but currently it's the easiest reactive proxy for a viewport resize event.
+    const contentWidth = viewportStore.useSelector(state => state.contentWidth)
+    const cursor = useSelector(state => state.cursor)
```
I want to check in about the potential performance implications of this. This would cause every visible `ThoughtAnnotation` instance to re-render whenever the cursor changes. Does this differe from `main`? Can this be narrowed down?

ethan-james commented on src/components/ThoughtAnnotation.tsx:
```diff
@@ -121,65 +119,56 @@ const ThoughtAnnotation = React.memo(
-      const labelId = findDescendant(state, head(simplePath), '=label')
-      const labelChild = anyChild(state, labelId || undefined)
-      return isEditing ? (liveValueIfEditing ?? value) : labelChild ? labelChild.value : value
-    })
```
#1419 looks OK.

raineorshine commented:
Thanks. I'm bringing the discussion here as there seems to be more at stake than originally anticipated. 

> If I remove the span, then it messes up the superscripts:
> 
> 
> 
> <img width="2400" height="600" alt="color-theme-superscript-on-light-theme-1-diff" src="https://github.com/user-attachments/assets/1b457f0f-bd9f-4524-9976-03752e6fd9ca" />
> 
> 
> 
> I thought it might be as simple as adding an explicit height to the container, such as `height: multiline ? '1.25em' : '2em'`. That did fix the superscripts, but it caused other layout problems:
> 
> 
> 
> <img width="2400" height="600" alt="color-theme-superscript-on-light-theme-1-diff" src="https://github.com/user-attachments/assets/c01023db-b5ed-4f9a-9f31-70266e09b7fd" />
> 
> 
> 
> I went ahead and removed the `textMarkup` and `placeholder` props so that it will only ever render `'&ZeroWidthSpace;'`. In terms of removing the span entirely, I think that there are two moving parts:
> 
> 
> 
> 1. Setting the height of the container correctly so that it leaves room for annotations without pushing lower thoughts out of place
> 
> 2. Figuring out if anything breaks now that we are relying on typographic CSS such as [verticalAlign: 'text-top'](https://github.com/ethan-james/em/blob/cf05215cf7deaf6674dcbb168b4391358177edd7/src/components/ThoughtAnnotation.tsx#L47) while no longer including any text in the annotation.
> 
> 


This is giving me some doubts about the current solution and whether I want to merge. The main point of this task was to remove the dependency on the the invisible placeholder element and reduce complexity. Now we have a hybrid solution which relies partly on DOM measurement and partly on the placeholder, which seems to me less good than either pure approach. 

We began with the theory that the superscript can be rendered a fixed number of em up from the bottom-right corner of the editable, without any placeholders. Is this not true? Maybe better understanding the path to true decoupling from the placeholder will help us get to the right solution.

ethan-james commented:
> > 1. Setting the height of the container correctly so that it leaves room for annotations without pushing lower thoughts out of place
> > 2. Figuring out if anything breaks now that we are relying on typographic CSS such as [verticalAlign: 'text-top'](https://github.com/ethan-james/em/blob/cf05215cf7deaf6674dcbb168b4391358177edd7/src/components/ThoughtAnnotation.tsx#L47) while no longer including any text in the annotation.
> 
> This is giving me some doubts about the current solution and whether I want to merge. The main point of this task was to remove the dependency on the the invisible placeholder element and reduce complexity. Now we have a hybrid solution which relies partly on DOM measurement and partly on the placeholder, which seems to me less good than either pure approach.
> 
> We began with the theory that the superscript can be rendered a fixed number of em up from the bottom-right corner of the editable, without any placeholders. Is this not true? Maybe better understanding the path to true decoupling from the placeholder will help us get to the right solution.

Maybe I've been treating `vertical-align` as more magical than it truly is. I removed to span to check things out, and nudging the superscript went pretty well. There's one table view test that's failing. I can look into that more tomorrow to see if it's related to `vertical-align`, or I can update the snapshot if you'd rather move forward with it as it is.

raineorshine commented:
The updated snapshot is okay with me!

raineorshine reviewed (changes requested):
Thanks!

There are a couple regressions related to Table View and Context View, which makes sense given the different layouts used there. Hopefully we can come up with a fairly general solution. I'll let you investigate and determine the effort involved.

I've added snapshot tests in ab158c293d to cover the cases that were missing, so you should be able to see the tests fail if you merge `main` now.

ethan-james commented:
> Thanks!
> 
> There are a couple regressions related to Table View and Context View, which makes sense given the different layouts used there. Hopefully we can come up with a fairly general solution. I'll let you investigate and determine the effort involved.
> 
> I've added snapshot tests in [ab158c2](https://github.com/cybersemics/em/commit/ab158c293dcbda1557e80e2000d3bdef740d21db) to cover the cases that were missing, so you should be able to see the tests fail if you merge `main` now.

Thanks for doing that, it was helpful for troubleshooting. I fixed the table view test completely, and I fixed the major issue with the context view where it was calculating the position of the thought annotation before the `disappearingUpperRight` animation was complete. That issue isn't reflected in the test, which instead shows small discrepancies that look similar to the other snapshots that we've updated.

I didn't see anything obviously different about the positioning of a context view thought compared to a regular thought, and indeed the superscript for `a` is off even thought it's not part of the active context view. I tried making small adjustments to the positioning [here](https://github.com/ethan-james/em/blob/5bc4502007543b14279fe9e936fd3b0c9a5b843e/src/components/ThoughtAnnotationWrapper.tsx#L31) or [here](https://github.com/ethan-james/em/blob/5bc4502007543b14279fe9e936fd3b0c9a5b843e/src/components/StaticSuperscript.tsx#L56), and ended up in a familiar situation where other tests start failing before the superscript nudges up by a pixel on `a`.

Anyway, I would recommend updating the snapshot for the context view test. Let me know if you want me to dig deeper instead.

raineorshine commented on src/components/ThoughtAnnotation.tsx:
```diff
@@ -279,13 +282,69 @@ const ThoughtAnnotationContainer = React.memo(
+            if (isInContextView && display !== undefined) annotationRef.current.style.display = display
+          }
+        },
+        isInContextView ? durations.get('disappearingUpperRight') : 0,
```
I found a couple cases where this delay doesn't work. 

One is if the context view is rapidly toggled:

https://github.com/user-attachments/assets/cf61b970-6701-4304-b0d4-285a6c5c1c7a

Another is if the cursor is quickly moved before the animation completes:

https://github.com/user-attachments/assets/5eb9cf27-206e-481e-9da1-37d85bf29646

raineorshine commented on src/components/ThoughtAnnotationWrapper.tsx:
```diff
@@ -13,41 +10,26 @@ import FauxCaret from './FauxCaret'
-        // maxWidth: '100%',
-        marginTop: '0',
+        lineHeight: multiline ? 1.25 : undefined,
+        marginTop: multiline ? 'calc(-0.12em - 0.5px) !important' : value ? '-0.425em' : undefined,
```
Let's avoid the use of `!important`. There should be other ways to set up the styles or create Panda recipes to allow style overriding in a less brute force way.

raineorshine reviewed (changes requested):
Thanks for the updat!

> Anyway, I would recommend updating the snapshot for the context view test. Let me know if you want me to dig deeper instead.

If the discrepancies are small and the position looks correct when using the app, I'm okay with updating the snapshots.

raineorshine commented:
Thanks! To thoroughly test the thought annotation, I would test superscripts, urls, code, and colors within normal view, table view, and context view.

Navigate around, resize the window, etc. This is not mobile specific so that should save you some time.

Note that there are some layout bugs related to table view in `main` so you'll need to compare the behavior with `main` to confirm if it's an actual regression or not.

Some possible test data: 
*(This can be copied and pasted directly into em)*
```
- 1
  - a
    - b
      - c
        - =view
          - Table
        - d
          - d1
        - e
          - e1
            - f
        - stackoverflow.com
          - stackoverflow.com
- 2
  - a
    - b
      - c
        - =view
          - Table
        - d
          - d1
        - e
          - e1
            - f
- 3
  - url
    - stackoverflow.com
  - code
```

raineorshine commented on src/hooks/useContextAnimation.ts:
```diff
@@ -0,0 +1,41 @@
+ * This is used to display the animation in TreeNode, and also to determine whether thought annotations should delay
+ * their position calculation until after the animation has completed (#3352).
+ */
+const useContextAnimation = (path: Path, isAppearing?: boolean) => {
```
This hook is too thin to justify the abstraction, as it is just a wrapper around `useSelector`. It would be better to factor out the selector and use `useSelector(contextAnimation(path, isAppearing))` as needed.

raineorshine commented on src/hooks/useContextAnimation.ts:
```diff
@@ -0,0 +1,41 @@
+    if (isDescendantPath(path, state.cursor)) {
+      const isCursorInContextView = isLastActionContextView
+        ? !!state.cursor && isContextViewActive(state, state.cursor)
+        : false
```
Literal `true` and `false` can always be reduced out of ternary expressions. This is equivalent to:

```ts
isLastActionContextView && !!state.cursor && isContextViewActive(state, state.cursor)
```

raineorshine commented on src/components/ThoughtAnnotation.tsx:
```diff
@@ -279,13 +284,78 @@ const ThoughtAnnotationContainer = React.memo(
     }, [])
 
+    const positionAnnotation = useCallback(() => {
+      if (editableRef.current && annotationRef.current) {
```
You can invert this condition and `return` to short circuit the function and make this a little cleaner by avoiding the extra indentation level.

raineorshine commented on src/components/ThoughtAnnotation.tsx:
```diff
@@ -279,13 +284,78 @@ const ThoughtAnnotationContainer = React.memo(
       setCalculateContexts(true)
     }, [])
 
+    const positionAnnotation = useCallback(() => {
```
This could use a header comment summarizing the purpose of this function.

It may also be worth it to pull `positionAnnotation` and the `useEffect` out into a hook in order to hide the implementation details from the rest of the component.

ethan-james commented on src/components/ThoughtAnnotationWrapper.tsx:
```diff
@@ -13,41 +10,26 @@ import FauxCaret from './FauxCaret'
-        // maxWidth: '100%',
-        marginTop: '0',
+        lineHeight: multiline ? 1.25 : undefined,
+        marginTop: multiline ? 'calc(-0.12em - 0.5px)' : value ? '-0.425em' : undefined,
```
For multiline, I consolidated the top margins from the children in ae2661a61c687b70ea4f55fa7a8a4c4d049f1ef9.

raineorshine commented:
Are you testing the sample data above? I'm seeing a few positioning issues:

<img width="999" height="471" alt="image" src="https://github.com/user-attachments/assets/4a9a5878-94be-4590-8c2a-52599f19dcb2" />

<img width="998" height="506" alt="image" src="https://github.com/user-attachments/assets/32e2e638-92d3-47e8-b42b-8f492fdccaee" />

ethan-james commented:
I think I fixed all of the table and double-annotation issues. The only thing left that I know of is the `code` issue. It turns out that [const textNode = editableRef.current.lastChild](https://github.com/ethan-james/em/blob/5732a9c844f9bbad77f4468b8698a821e9e63130/src/hooks/usePositionedAnnotation.ts#L36) stops working when a style is applied and a `font` tag is inserted. It works OK if the style is only applied to the first part of the thought, but a thought with a style applied to all of it no longer has a `textNode` to reference as a direct child of `editableRef.current`. Do you want me to keep going and make that logic more robust in order to solve what is hopefully the last positioning issue, or do you think we should re-evaluate?

raineorshine commented:
Thanks for the update!

Yes, we will need to handle nested HTML. It may also occur with `<b>`, `<i>`, etc as well.

`editableRef.current.lastChild` seems a bit fragile since it is tightly coupled not just to the main DOM node of a different component, but to the structure of its descendants. I would be nice if there was a way to decouple that, but we can focus on behavioral integrity before we consider further refactoring.

raineorshine reviewed (changes requested):
Thanks for giving some thought to the coupling with `editableRef`. Looking at the new helper function `getFinalTextNode`, I think I'm leanings towards keeping the existing direct reference rather than trying to decouple it. None of the refactoring options are super attractive. Importantly, `getFinalTextNode` is pretty general. While there is a tight coupling with the descendant text node, it's not like it relies on a specific DOM structure. It just assumes that there is a text node *somewhere* within `editableRef`. I think this is okay. Regrettably, there are a lot of components that depend on `editableRef` directly so we don't have great separation of concerns between the different thought wrappers, but that is what it is.

Detailed testing coming soon.

ethan-james commented on src/components/StaticSuperscript.tsx:
```diff
@@ -52,7 +56,7 @@ const StaticSuperscript = React.forwardRef<
                 position: 'relative',
                 zIndex: 'stack',
-                top: '-1px',
+                top: type === 'multiline' ? '-2em' : type === 'singleline' ? '-2.745em' : undefined,
```
I'm actually going to push my changes so that we can make sure it looks the same when the CI test suite runs.

ethan-james replied:
I was able to fix the tests by re-introducing some typography (a zero-width space) back into the annotation element. It's hard to say exactly why this is necessary because it looked fine everywhere except inside Puppeteer. I tried replacing the character with `height: multiline ? '1.25em' : '2em'` to make the annotation wrapper occupy the required amount of space, but that didn't fix the problem and it did break the layout in other tests. If you'd like, I can take another look tomorrow to see why typography is still so important to the layout of this element, or I can wrap the zero-width space so that it only shows up in CI.

There was one 13px superscript test that was slightly off after moving the positioning into a wrapper around `StaticSuperscript`, but I went ahead and updated that.

ethan-james commented:
I think I've managed to fix the incorrect positioning, but unfortunately the fix creates a bit of a flicker in some situations. A better solution would be to remove `setTimeout` from normal (non-animated) annotations, and that almost works but it seems to cause problems with multiline thoughts, especially when a multiline URL becomes the cursor thought and goes from ellipsized to un-ellipsized.

I thought maybe I could watch for changes to `multiline` and treat that as a special case, but I'm not sure it would solve the flicker, and it didn't work anyway. When a URL goes from ellipsized to un-ellipsized (multiline) it is measured [asynchronously](https://github.com/ethan-james/em/blob/56ccf65ada7c01193f08cba8f9c3b9e48222f793/src/components/Editable/useMultiline.ts#L43) and is not updated until it's too late to update positioning. I'll let you know if I can think of a better solution.

ethan-james commented:
Actually, I realized that I can use `multiline` as a dependency to trigger an async re-positioning for multiline thoughts, eliminating the need for a timeout if the thought isn't multiline. Seems to work OK so far!

ethan-james commented:
> I can confirm that [Edited thought superscript position incorrect on first paint](https://github.com/cybersemics/em/pull/3357?notification_referrer_id=NT_kwDOAAtyxLIyMDI3ODI5ODgxNzo3NTAyNzY#pullrequestreview-3602545278) is fixed.
> 
> I'm still seeing the broken behavior after toggle context view: [Deactivated context view superscript position incorrect on first paint](https://github.com/cybersemics/em/pull/3357?notification_referrer_id=NT_kwDOAAtyxLIyMDI3ODI5ODgxNzo3NTAyNzY#issuecomment-3680330877).
> 
> The test case from [#3357 (comment)](https://github.com/cybersemics/em/pull/3357#issuecomment-3620946333) still looks solid.

It seems like when virtualized thoughts re-enter the viewport after the context view thoughts leave, they come back with the annotation in the wrong position. I'm not sure why that would be the case, but hiding the annotation until positioning occurs seems to reliably fix the issue.

fbmcipher commented:
@raineorshine Sure thing – I'll run a couple quick tests and post my results here tomorrow.

ethan-james commented:
> 1. It looks like some superscripts are being hidden that don't need to be hidden. It should only affect children that are being animated in/out, not the cursor thought, siblings, or ancestors.

It seems like the context animation is applied to all thoughts, not just ones that are entering or leaving the context view. Probably because it's tough to determine which thoughts are animating out and leaving the DOM. Do you think it's worthwhile to try to narrow that down so that the animation is not applied to thoughts that are higher up than the cursor thought?

ethan-james commented:
> > 1. It looks like some superscripts are being hidden that don't need to be hidden. It should only affect children that are being animated in/out, not the cursor thought, siblings, or ancestors.
> 
> It seems like the context animation is applied to all thoughts, not just ones that are entering or leaving the context view. Probably because it's tough to determine which thoughts are animating out and leaving the DOM. Do you think it's worthwhile to try to narrow that down so that the animation is not applied to thoughts that are higher up than the cursor thought?

It seems like using [belowCursor](https://github.com/ethan-james/em/blob/1740231c9f67c71fd2157df650ea5fb232c1d97a/src/selectors/linearizeTree.ts#L120) works pretty well to narrow down which annotations are recalculated, apart from some unpleasant prop drilling. However, there are still thoughts down below that shouldn't recalculate, as you can see if you start from the repro in https://github.com/cybersemics/em/pull/3357#issuecomment-3680330877 and toggle the context view on `c` instead of `b`.

https://github.com/user-attachments/assets/ecfbfe37-ce30-405d-a386-cafdf13fe4b8

raineorshine reviewed (commented):
After playing with this a bit, I unfortunately have come to the conclusion that rendering the superscripts after a delay is too disruptive to the UX. This is true whether the superscripts snap in or fade in smoothly. 

Theoretically it's possible to measure the bounding box synchronously in `useLayoutEffect` and render the superscript before the first paint, though it will have a performance impact. I'd love to get your input on how viable this might be before we spend additional time.

I did another review below, but they are moot points if we're unable to efficiently render the superscript synchronously.

Thanks so much for your work.

---

I'm still seeing the issue described at the end of this comment. Are you able to reproduce? https://github.com/cybersemics/em/pull/3357#pullrequestreview-3619256831

> **Note 1:** This was my interpretation of the test based on the instructions given and the fact that the `positionAnnotation` callback executes after the Context View animation completes. Let me know if I got it wrong.

I don't think this changes the results necessarily, but I meant a simple cursor navigation in normal view that expands a context. Activating the Context View has its own performance issues, so I would avoid that and test superscript rendering without the context view animation. Sorry for the confusion!

> * Measure the **time from tapping the Context View icon** to **the point where layout fully settles**, using frame-by-frame timings and snapshots taken by the Chrome Performance Profiler.

I'm not sure that measuring the time in between the start and finish really measures performance. That's just measuring animation duration, which should generally be constant. If performance drops, the browser will drop frames without changing the animation duration. Only if the CPU is thrashing will the animation actually slow down. I think we need to be measuring either the CPU usage in ms or the average frame rate, ideally without the confounding variable of the Context View animation.

raineorshine commented:
Oops, I meant selection range, not bounding box! The question is whether it's possible or not to use the same technique that we are using here, but before the first paint.
